### DISCUSSION TITLE
Pobieranie

### DISCUSSION THREAD

**Discussion by kkuriata** [TowarzyszFatCat/doccli#20] [Ideas]
Być może jest szansa na dodanie funkcji masowego pobrania anime w danej rozdzielczości? Korzystam z apki, ale często wyjeżdżam i wtedy ściągam sobie rzeczy na trasę. Niestety żmudne pobieranie po 1 odcinku jest słabe, fajnie by było, jakby dało się to robić w łatwy sposób masowo.

**Comment by TowarzyszFatCat** [comment]
Tak, planuję niebawem coś takiego dodać. Tu mniej więcej opisałem jak to widzę:
https://github.com/TowarzyszFatCat/doccli/discussions/13#discussioncomment-10739387

**Comment by mareczek82** [comment]
to w tej chwili jest opcja pobierania pojedynczych epizodów? gdzie ją znajdę?

**Reply by TowarzyszFatCat** [reply]
Nie ma, nie robiłem nic przy projekcie już prawie pół roku. Będę przepisywał doccli jeszcze raz jak znajdę trochę czasu. Jednak problemem jest to, że nowe stronki z anime i te starsze mają coraz to lepsze zabezpieczenia i nie da się nic z nich pobrać. Wsm przepisywanie doccli zacząłem już trochę temu i nawet ładnie mi to wyszło, ale co mi z pisania takiej aplikacji jak nie ma źródeł.

Planuję wznowić projekt, ale najpierw muszę znaleźć lepszy sposób na pozyskiwanie źródeł bo jedynym pewnym źródłem jest CDA od docchi, z resztą jest problem.

**Reply by TowarzyszFatCat** [reply]
@mareczek82 Dodałem funkcje pobierania na tyle na ile mogłem i od razu mówię, że nie będzie działać z każdym źródłem. Aktualny kod tylko na tyle mi pozwolił, z pobieraniem bulk kilku odcinków na raz będzie problem przy obecnym api.

**Reply by mareczek82** [reply]
@TowarzyszFatCat dziekuje! 
Ja chyba źle zrozumiałem @kkuriata, bo myślałem, że juz ściąganie było tylko, że pojedynczych odcinków a ja to przeoczyłem.

Dzięki za update, zaraz przetestuję!

**Reply by mareczek82** [reply]
nie wiem czy to pomoże, ale jest taki dodatek do stremio i on ładnie "wczytuje" źródła. https://github.com/skoruppa/docchi-stremio-addon

**Reply by TowarzyszFatCat** [reply]
Widziałem ten projekt, ich api jest bardzo podobne do mojego, ale też nie obsługuje mega.nz. W doccli jest zrobione na to obejście za pomocą megatools - pobiera plik do /tmp i odtwarza, ale i tak wszystko zależy od ustawień prywatności wideo, bo tego już nie obejdę.

Przy obecnym api musiałbym na każdą konkretną strone/player pisać jakieś obejście, ale to jest bez sensu, patrząc na to, że te playery dostają co trochę jakiś update i często już tak miałem, że musiałem zmieniać fragmenty kodu bo nie chciało z nich oddtwarzać. Teraz opieram wszystko na yt-dlp, ale też nie radzi sobie z niektórymi playerami. Dodatkowo scrapper od docchi, z którego teraz korzystam do pobierania danych o anime i playerów jest napisany w TypeScript i nie używam go bezpośrednio, tylko łączę się z api strony internetowej, co nie jest optymalne. 

Jest dużo rzeczy, przez które odechciało mi się pracować nad projektem, bo aktualizowanie tego kodu sprawia dużo problemów. Zabierałem się za kompletny rewrite z bardziej dopracowanym gui i integracją z anilistą [[ten branch]](https://github.com/TowarzyszFatCat/doccli/tree/v3-rewrite) ale zaczęły się problemy z paczkami których używałem i kilka z nich przestało być wspieranych. Teraz dopisałem te kilka linijek kodu, żeby dało się cokolwiek pobierać, bo aż się zdziwiłem, że ktokolwiek jeszcze zagląda w to repo hah.

W przyszłości planuję zrobić z tego aplikację, bo terminal jednak ma swoje ograniczenia. Wiele problemów miałem też z paczkami bo w każdym terminalu wyświetlanie działało lekko inaczej, a na linuxie jest ich multum i min. dlatego porzuciłem ten rewrite. W skrócie chciałem wycisnąć wszystko co się da, ale trochę mnie to przerosło, bo więcej czasu spędzałem na naprawianiu paczek których używam niż nad swoim własnym kodem.

Dzięki @mareczek82, że wciąż wspierasz projekt i w ogóle tu zaglądasz, bo ostatni update był jak nie patrzeć ponad 10 miesięcy temu.
W sumie wszystkim innym też dziękuję za wsparcie i zainteresowanie, bo to daje motywację, żeby kiedyś to wszystko porządnie ogarnąć.

**Reply by mareczek82** [reply]
Hej, 

przepraszam, umknęła mi ta wiadomość. dziękuję za rytualizację!

**Comment by TowarzyszFatCat** [comment]
Cześć @kkuriata i @mareczek82. Wracam do tematu. Trochę musieliście na to poczekać, ale w końcu znalazł się czas i wróciły chęci, więc wszystko o co prosiliście zostało już dodane do doccli.

Udało się zaimplementować masowe pobieranie sezonów (niestety bez wyboru jakości, więc program pobiera zawsze w najlepszej dostępnej), Zrobiłem też w końcu pełne wsparcie na Windowsa. Dodatkowo, walcząc z tymi odtwarzaczami, udało mi się dorzucić sprawdzanie źródeł na żywo (program teraz sam przewiduje i weryfikuje w tle, który link faktycznie zadziała)

Dzięki za cierpliwość i mam nadzieję, że mimo tych wszystkich przerw nadal używacie programu. Jestem niezmiernie ciekaw, czy macie jakieś inne pomysły na rozwój i czego wam jeszcze brakuje na ten moment.

**Reply by mareczek82** [reply]
Witaj Towarzyszu,

niestety mam problem z instalacją, szczegóły w oddzielnym issue. 

Super, że wróciłeś!

**Reply by mareczek82** [reply]
Jestem po pierwszych testach. 

Na pierwszy ogień poszło kill la kill. 
- pobrało całość
- po wymuszeniu zamknięcia ctrl+c i ponownym włączeniu apki **pomija pobrane już odcinki** i **_wznawia rozpoczęte od miejsca gdzie przerwałem!!!_**

Czego mi brakuje 
- usunięta opcja pobierania pojedynczego odcinka
- wyboru opcji jakości - ale to już mniej, zawsze mogę handbraka sobie puścić

Super, że wróciłeś do projektu!

**Reply by TowarzyszFatCat** [reply]
Akurat wypuściłem kolejny ogromy update, pobieranie pojedyńcze przywrócę jeszcze dzisiaj w jakimś hotfixie bo to bez problemu można ogarnąć. Z wyborem jakości zobaczymy co da się zrobić ale cudów nie obiecuje :P

**Reply by TowarzyszFatCat** [reply]
Wszystko o co prosiłeś udało się dodać :) [v2.30.1](https://github.com/TowarzyszFatCat/doccli/releases/tag/v2.30.1)
